### DISCUSSION TITLE
Feedback on rbyte

### DISCUSSION THREAD

**Discussion by sandhawalia** [foxglove/mcap#1242] [General]
Hey folks,

Thanks for the amazing work you’ve done on mcap 🙇🏽. At [Yaak](https://www.yaak.ai/) we are building workflows for training end-to-end AI for robotics. Towards that we’ve open-sourced [rbyte](https://github.com/yaak-ai/rbyte). With the [latest release](https://github.com/yaak-ai/rbyte/releases/tag/v0.2.0) we've added support for building training batches (PyTorch tensordict) directly from a list of mcap files(s).

Would be great to get your opinion and feedback on [rbyte](https://github.com/yaak-ai/rbyte). A sample [NuScenes-v1.0](https://github.com/yaak-ai/rbyte/blob/main/examples/config_templates/dataset/mcap_protobuf.yaml) config is in the repo for you to try out. If there public robotics datasets in mcap, which you would like us to support in rbyte, please feel free ping us.

**Comment by jtbandes** [comment]
Thanks for sharing! Have not dug in too much yet, but I'm curious why you need so much custom mcap reading code in McapFrameReader? Is there something about the format itself or the mcap python library that is not suiting your needs?

**Reply by sandhawalia** [reply]
Hi there, thanks for having a look. Paging @egorchakov for your question around McapFrameReader.

**Reply by egorchakov** [reply]
Hey @jtbandes, thanks for taking a look! 

I'm new to `mcap` so it's totally possible I missed/misunderstood something about the structure. 

With `McapFrameReader`, we want to be able to read the `i`-th frame on the specified topic in a random access pattern. AFAICT (correct me if I'm wrong here), there's no  _global_ message index available in the summary section, so we instead build one using [Chunk Index](https://mcap.dev/spec#chunk-index-op0x08) and the corresponding (intra-chunk) [Message Index](https://mcap.dev/spec#message-index-op0x07). For each frame index `i`, we store the corresponding chunk index, message start offset and message length. The logic for getting the `i`-th frame essentially boils down to seeking the chunk containing the `i`-th message and then reading the message using offset and length.

If you have ideas on simplifying this I'd love to hear them!
